pkg/controller/persistentvolume/persistentvolume_recycler_controller.go

```diff
@@ -40,21 +40,39 @@ var _ volume.VolumeHost = &PersistentVolumeRecycler{}
 // This controller will Recycle those volumes whose reclaim policy is set to PersistentVolumeReclaimRecycle and make them
 // available again for a new claim.
 type PersistentVolumeRecycler struct {
-	volumeController *framework.Controller
-	stopChannel      chan struct{}
-	client           recyclerClient
-	kubeClient       client.Interface
-	pluginMgr        volume.VolumePluginMgr
-	cloud            cloudprovider.Interface
+	volumeController  *framework.Controller
+	stopChannel       chan struct{}
+	client            recyclerClient
+	kubeClient        client.Interface
+	pluginMgr         volume.VolumePluginMgr
+	cloud             cloudprovider.Interface
+	maximumRetryCount int
+	syncPeriod        time.Duration
+	// Local cache of failed recycle / delete operations. Map volume.Name -> status of the volume.
+	// Only PVs in Released state have an entry here.
+	releasedVolumes map[string]releasedVolumeStatus
```

Comment:
See buried comment above https://github.com/kubernetes/kubernetes/pull/19365#discussion_r49107689

Using PV annotations generated many events as the PV was saved over and over. Also require more keys.  Code was smaller after switching to local state.  In the case of controller restart, all PVs start at 0, so the MaxRetries var is not guaranteed (it would be w/ PV annotation), but that doesn't seem important.